### ISSUE TITLE
关于在 CSI300 上复现实验结果的问题

### ISSUE THREAD

**Issue by summerd65** [BerkinChen/AlphaSAGE#4]
您好，
我在 CSI300 上尝试了多次实验，目前得到的 IC 值依然在 0.05 左右，还没有达到论文中提到的 0.079。
下面是我目前复现实验的完整流程，想请您帮忙看看是否有哪里设置不太对：

1️⃣ 下载 Qlib
`git clone https://github.com/microsoft/qlib.git && cd qlib`

2️⃣ 准备基础数据
`python scripts/get_data.py qlib_data --target_dir ~/AlphaSAGE/data/qlib_data/cn_data --region cn`

3️⃣ 生成复现实验所需的数据（cn_data_rolling）
`cd ~/AlphaSAGE/src/data_collection && python fetch_baostock_data.py`

4️⃣ 生成 alpha pool（在 train_gfn.py 中把 QLIB_PATH 改成 'data/qlib_data/cn_data_rolling'）
`python train_gfn.py \
    --seed 0 \
    --instrument csi300 \
    --pool_capacity 50 \
    --log_freq 500 \
    --update_freq 64 \
    --n_episodes 10000 \
    --encoder_type gnn \
    --entropy_coef 0.01 \
    --entropy_temperature 1.0 \
    --mask_dropout_prob 1.0 \
    --ssl_weight 1.0 \
    --nov_weight 0.3 \
    --weight_decay_type linear \
    --final_weight_ratio 0.0`

（这是生成完 alpha pool 的结果截图：）

<img width="1546" height="907" alt="Image" src="https://github.com/user-attachments/assets/5247fc2a-a221-478b-8037-93140ba80b1f" />

5️⃣ 合并 alpha pool（在 run_adaptive_combination.py 中把 QLIB_PATH 改成 'data/qlib_data/cn_data_rolling'）
`python run_adaptive_combination.py \
    --expressions_file data/gfn_logs/pool_50/gfn_gnn_csi300_50_0-0.01-1.0-1.0-1.0-0.3-linear-0.0/pool_9999.json \
    --instruments csi300 \
    --threshold_ric 0.015 \
    --threshold_ricir 0.15 \
    --chunk_size 400 \
    --window inf \
    --n_factors 20 \
    --cuda 2 \
    --train_end_year 2020 \
    --seed 0`

（这是最后的结果截图：）

<img width="906" height="230" alt="Image" src="https://github.com/user-attachments/assets/e4c5afc9-0a6d-430a-87be-db25efdc9ec6" />

想请问：
	•	我目前数据准备和复现实验的流程是否正确？
	•	训练出来 IC 一直在 0.05 左右，这种情况正常吗？是否可能是数据或参数的原因？

再次感谢您开源这个非常棒的项目，也非常期待能得到您的建议 ，谢谢！🙏

**Comment by summerd65** [comment]
您论文当中对 CSI300 的训练/测试数据时间是从 2010-01-01 到 2020-12-31 / 2022-01-01 到 2024-12-31，但是 train_gfn.py 里对训练/测试数据的时间段划分不同：

<img width="863" height="64" alt="Image" src="https://github.com/user-attachments/assets/2d5ed6af-aded-436b-b9d3-b89430e1c26b" />

我尝试使用论文当中的划分方法，在 CSI300 上的结果如下：
1. 生成 alpha pool：

<img width="1462" height="856" alt="Image" src="https://github.com/user-attachments/assets/1abdd335-bc60-4921-865b-51cd4cc1330d" />

2. 合并 alpha pool：

<img width="1243" height="293" alt="Image" src="https://github.com/user-attachments/assets/7a6eb55b-d61a-4ae4-a230-fec40f5fa90d" />

最后得到的 IC 在 0.06。我也尝试了其它 seed，还是无法得到 0.079 的 IC (seed 取 42 时得到的 IC 只有 0.0367)。不知道是否因为我的数据准备有误，还是生成 alpha pool 的超参数没有设置正确，十分希望能得到您的建议，谢谢！

我在运行程序的过程中，只遇到过两个 warnings：
1. 有关 TensorFlow 的：

<img width="1269" height="97" alt="Image" src="https://github.com/user-attachments/assets/2b2d810b-a296-49fc-9498-c4ab942a531e" />

2. 有关 torch 里 lr_scheduler 的：

<img width="1150" height="116" alt="Image" src="https://github.com/user-attachments/assets/c625abf3-99f4-4c1e-8db1-6a497ed71c87" />

**Comment by langslike** [comment]
alphaforge的效果和这个alphasage比，哪个好点？ @summerd65

**Comment by summerd65** [comment]
alphaforge 我没尝试过，alphasage 我最后也没跑出论文中的结果，不知道问题出在哪里 @langslike

**Comment by langslike** [comment]
能跑到0.05也算不错了， 你如果做量化的话， 可以交流交流，我微信kurtosisss @summerd65

**Comment by TangSuger7** [comment]
> 它的方法应该是注重因子组合的效果而不是单因子的 [@sdjafkhk](https://github.com/sdjafkhk)

挖出来的因子单个信息因子并不高，而且不稳定，也就是说最终还依赖多因子的动态加权，那为什么不用其他方法挖出ic更高的再动态加权呢

**Comment by royaldragon** [comment]
我也被困在这个问题很久，最终终于找到是  fetch_baostock_data.py 的问题。 由于作者没有放上训练时的数据集，也没有说明如何准备数据， 我看大家一般都是通过 Qlib来下载数据（只到2022年），然后用 baostock 来补充数据（但这里有bug，所以作者很可能不是通过这种方式来下数据的，至于怎么拿到2022年后的数据，论文里也没说），恰恰是 fetch_baostock_data.py里有bug，所以算前复权因子取错基准行、价格序列被扭曲，所以怎么训都得不到论文的结果。

## Bug：`as_of_date` 在 RangeIndex 上 searchsorted 日期，导致前复权因子取错基准行、价格序列被扭曲

### 影响范围

- 所有使用 `foreAdjustFactor`（前复权）模式构建 Qlib 数据集的用户
- 继承了 alphagen 系 `fetch_baostock_data.py` 的项目（AlphaForge、AlphaSAGE等）
- 使用 `backAdjustFactor`（后复权）模式的用户**不受影响**

### Bug 位置

`fetch_baostock_data.py` 的 `as_of_date` 函数，约第 178 行（buggy 上游版本）：

```python
# BUGGY（上游 alphagen 原版）
def as_of_date(df: pd.DataFrame, date: str) -> pd.Series:
    index: int = df.index.searchsorted(date, side="right") - 1
    return df.iloc[index]
```

### 根因
pd.DataFrame(lst, columns=res.fields) 创建的 DataFrame 的 index 是默认的 RangeIndex（0, 1, 2, ...），不是 DatetimeIndex。searchsorted("2024-12-31") 把日期字符串和整数做比较，Python 不会报错（字符串和整数比较返回一致但语义无意义的排序），静默返回了错误的行号。

后果
前复权模式下，as_of_date 的返回值被用于计算归一化基准：
ref = 1.0 / foreAdjustFactor["as_of_date"]
price *= ref
foreAdjustFactor *= ref

取错基准行 → 每条记录的 ref 值不对 → 价格序列和复权因子序列的形状被整体扭曲（不是常数缩放，而是随时间变化的畸变）。直观表现：浦发银行 $factor@2024-12-31 应该是 0.9268，实际变成了 7.68。

修复

### FIXED
def as_of_date(df: pd.DataFrame, date: str) -> pd.Series:
    index: int = df["date"].searchsorted(date, side="right") - 1
    return df.iloc[index]


### Bug 继承链
该 bug 来源于 alphagen（https://github.com/ICT-FinD-Lab/alphagen）的 fetch_baostock_data.py（AlphaGen 官方仓库），经 AlphaForge 继承到 AlphaSAGE/TemporalAlpha。
在 alphagen 源头，默认使用 backAdjustFactor（后复权，adjustflag="1"）且 adjust_date="2009-01-01" 固定。此时 bug 只造成常数缩放，对收益率计算无影响，因此长期未被发现。
当 AlphaSAGE 改为 foreAdjustFactor（前复权，adjustflag="2"）且 forward_adjust_date=today 后，bug 被引爆——价格序列形状被扭曲而非简单缩放。

对 Alpha 挖掘 IC 的影响（CSI300，2-seed 均值）
• 修复前：0.0666
• 修复后：0.0784

<img width="2278" height="190" alt="Image" src="https://github.com/user-attachments/assets/f77a7123-6b79-42f5-8d5f-fa1eeef45d09" />

<img width="2446" height="170" alt="Image" src="https://github.com/user-attachments/assets/ef05bd8a-a1bf-48b7-9753-a61b2c7dbfe3" />

其实作者如果放上数据集，就不会有这个问题，但如果要自己构建数据集，用到 fetch_baostock_data.py 就会有问题了。 或者现在也有社区版的 Qlib 数据，能覆盖2010 - 2024 的数据，但这个没试过。

**Comment by Fuzzytariy** [comment]
> 您论文当中对 CSI300 的训练/测试数据时间是从 2010-01-01 到 2020-12-31 / 2022-01-01 到 2024-12-31，但是 train_gfn.py 里对训练/测试数据的时间段划分不同：
> 
> <img width="863" height="64" alt="Image" src="https://github.com/user-attachments/assets/2d5ed6af-aded-436b-b9d3-b89430e1c26b" />
> 
> 我尝试使用论文当中的划分方法，在 CSI300 上的结果如下：
> 1. 生成 alpha pool：
> 
> <img width="1462" height="856" alt="Image" src="https://github.com/user-attachments/assets/1abdd335-bc60-4921-865b-51cd4cc1330d" />
> 
> 2. 合并 alpha pool：
> 
> <img width="1243" height="293" alt="Image" src="https://github.com/user-attachments/assets/7a6eb55b-d61a-4ae4-a230-fec40f5fa90d" />
> 
> 最后得到的 IC 在 0.06。我也尝试了其它 seed，还是无法得到 0.079 的 IC (seed 取 42 时得到的 IC 只有 0.0367)。不知道是否因为我的数据准备有误，还是生成 alpha pool 的超参数没有设置正确，十分希望能得到您的建议，谢谢！

请教下，修复fetch_baostock后您这边复现成功了吗，我这边修复后测出来的ic还是在0.05-0.06左右
